### DISCUSSION TITLE
Pickled assets on serverless are not accessible

### DISCUSSION THREAD

**Discussion by gnilrets** [dagster-io/dagster#17965] [Q&A] [answered]
We're using dagster serverless.  I developed an asset pipeline locally and got everything working.  I have a few assets that return some small simple Python classes that are then used as inputs in other assets.  These assets are just pointers to Snowflake objects.  Locally, I could materialize the assets independently without issue.  However, in serverless, it looks like the materialized assets are not accessible because they're pickled in a `/tmp` directory.

What is the right/easy way to persist these simple objects in serverless?  Do I need to set up a cloud storage integration or is there something easier for small materialized assets?

**Comment by prha** [comment]
Are you binding an IO manager explicitly?  The IO manager in serverless should default to a cloud-backed serverless specific IO manager rather than the local filesystem-based one.

If you are explicitly binding an IO manager to your jobs, you may want to bind it to the following:
```
from dagster_cloud.serverless.io_manager import serverless_io_manager

...
defs = Definitions(
    ...
    resources={"io_manager": serverless_io_manager}
)
```

**Reply by gnilrets** [reply]
No, I'm not explicitly using any IO Manager

**Reply by prha** [reply]
Hmm, it should be using the serverless IO manager then... can you confirm the type of io manager from within the asset?

```
@asset
def my_asset(context):
    context.log.info(f"iomanager: {type(context.resources.io_manager)}")
    ...
```

Also, the default serverless IO manager has been around for a while now... can you confirm the version of `dagster-cloud` you have packaged with your code location?

**Reply by gnilrets** [reply]
Ok, I tested out that log message on a branch deployment.  I think this means it's NOT using the serverless io manager?
```
iomanager: <class 'dagster._core.storage.fs_io_manager.PickledObjectFilesystemIOManager'>
```


When I go to view the dagster library versions in the UI, this is what I see:

<img width="1374" alt="image" src="https://github.com/dagster-io/dagster/assets/2371577/0fe6cd62-4399-4552-b9be-bf747759a66c">

There's no `dagster-cloud` listed.  However, I do install dagster-cloud in my requirements.txt file:
```
dagster==1.4.17
dagster-cloud==1.4.17
dagster-cloud-cli==1.4.17
dagster-dbt==0.20.17
dagster-ext-process==1.4.17
dagster-graphql==1.4.17
dagster-shell==0.20.17
dagster-slack==0.20.17
dagster-webserver==1.4.17
```

**Reply by prha** [reply]
We do fall back to the PickledObjectFilesystemIOManager when we encounter an error instantiating the serverless IO manager.

Can you confirm the values you get back from:
```
module_name = os.getenv("DAGSTER_DEFAULT_IO_MANAGER_MODULE")
attribute_name = os.getenv("DAGSTER_DEFAULT_IO_MANAGER_ATTRIBUTE")
```

If those are not mapping to `dagster_cloud.serverless` and `serverless_io_manager` respectively, please flag that.

Also, I wonder if you could get an explicit stack trace if you tried importing and binding the IO manager explicitly (in the definitions object).

**Reply by gnilrets** [reply]
Ok, so with this asset defined
```python

@asset
def my_asset(context):
    """This is just a test."""

    context.log.info(f"iomanager: {type(context.resources.io_manager)}")

    module_name = os.getenv("DAGSTER_DEFAULT_IO_MANAGER_MODULE")
    attribute_name = os.getenv("DAGSTER_DEFAULT_IO_MANAGER_ATTRIBUTE")
    context.log.info(f"module_name: {module_name}")
    context.log.info(f"attribute_name: {attribute_name}")
```

I get this log message:
```
iomanager: <class 'dagster._core.storage.fs_io_manager.PickledObjectFilesystemIOManager'>
module_name: dagster_cloud.serverless
attribute_name: serverless_io_manager
```


I also tried to import the IO manager explicitly but I got a message that there is no module named `boto3`.  I don't use S3 so it's not in my container.  I'll try rebuilding it with boto3...

**Reply by gnilrets** [reply]
After installing boto3 and spinning up a local version, I get this error message:
```
dagster._core.errors.DagsterInvalidDefinitionError: Conflicting versions of resource with key 'io_manager' were provided to different assets. When constructing a job, all resource definitions provided to assets must match by reference equality for a given key.
```

This is confusing because the only reference to an `io_manager` key in my entire code base is the one I just defined to test this out.

**Reply by gnilrets** [reply]
Ok, I think I found out what's doing this: dbt-cloud!  I defined a dbt cloud asset like this:
```python
import os

from dagster import EnvVar
from dagster_dbt import DbtCloudClientResource, load_assets_from_dbt_cloud_job

from eta_wh.resources import ENV_NAME

base_views_assets = load_assets_from_dbt_cloud_job(
    dbt_cloud=DbtCloudClientResource(  # type: ignore[arg-type]
        auth_token=EnvVar("DBT_CLOUD_API_TOKEN"),
        account_id=int(os.getenv("DBT_CLOUD_ACCOUNT_ID", "0")),
    ),
    # BuildBaseViews jobs
    job_id=1,
    node_info_to_group_fn=lambda node: "dbt_cloud",
)

dbt_build_base_views_job = define_asset_job(
    "dbt_build_base_views", selection=AssetSelection.groups("dbt_cloud")
)
```

When I include these assets in the defs, the io manager is `iomanager: <class 'dagster._core.storage.fs_io_manager.PickledObjectFilesystemIOManager'>`.  However, if I exclude the dbt-cloud assets from defs, without making any other changes or setting the io_manager explicitly to serverless, I get `iomanager: <class 'dagster_cloud.serverless.io_manager.PickledObjectServerlessIOManager'>`.

I've observed other issues with dbt cloud (https://github.com/dagster-io/dagster/issues/17931, https://github.com/dagster-io/dagster/issues/16726).

Any suggestions on how I could go about fixing this without having to get rid of dbt-cloud?

**Reply by prha** [reply]
@gnilrets are you using PEX, or the fast deploy CI/CD integration?  I'm wondering if there's some early binding happening when we build the PEX image.  We could confirm that if you see the serverless io manager when you turn off fast deploys.

**Reply by gnilrets** [reply]
@prha - Do you think that the deployment method would affect the io manager for dbt cloud?  That seems to be what's messing things up here.  If I remove dbt cloud job from my defs, it uses the serverless io manager by default.  When I include the dbt cloud job, it uses the filesystem io manager.

Either way, I don't think I'm using PEX because I've never heard of that before.  I'm building a custom image and deploying as follows (in gitlab CI):

```yaml
    # Build the dagster cloud container
    - DOCKER_BUILDKIT=1 dagster-cloud serverless build
      --base-image=eta-dagster-app
      --deployment=${DEPLOYMENT_ID}
      --source-directory=.
      ${CI_COMMIT_REF_NAME}

    # Get dagster cloud registry info
    - dagster-cloud serverless registry-info
        --organization=${DAGSTER_CLOUD_ORGANIZATION}
        --api-token=${DAGSTER_CLOUD_API_TOKEN}
        --deployment=${DEPLOYMENT_ID}
      > registry_info.env
    - source registry_info.env

    - dagster-cloud workspace add-location
      --organization=${DAGSTER_CLOUD_ORGANIZATION}
      --api-token=${DAGSTER_CLOUD_API_TOKEN}
      --deployment=${DEPLOYMENT_ID}
      --package-name=eta_wh
      --image=${REGISTRY_URL}:${CI_COMMIT_REF_NAME}
      mainspring_warehouse
```

**Reply by prha** [reply]
@gnilrets I've tried adding dbt cloud assets to a test project on serverless and still am not able to repro.  I'm wondering if the culprit might be 
A) if there's something specific about how you're building your code location images 
B) there's some bug in the cacheable dbt cloud assets (are you using cacheable job definitions as outlined in https://docs.dagster.io/integrations/dbt-cloud#step-4-cache-the-dbt-cloud-job-compilation?).

For A) how are you uploading the built image (e.g. before calling the `add-location` command?  The latest versions of the gitlab CI script makes uses of the `dagster-cloud serverless deploy` command: https://github.com/dagster-io/dagster-cloud-action/blob/main/gitlab/serverless-ci.yml

For B) If you set up job caching, I wonder if it makes a difference to manually set the `DAGSTER_DEFAULT_IO_MANAGER_MODULE` / `DAGSTER_DEFAULT_IO_MANAGER_ATTRIBUTE` variables in your CI environment.

**Reply by gnilrets** [reply, marked as answer]
Oh hey!  I was was changing too many things at once.  I think I've finally pinpointed the issue.  The dbt cloud thing was just a red herring.

It was just the fact that I didn't have the `boto3` library.  If I reset back to where I started and just add that package in my requirements file, it correctly uses the local filesystem io manager when run locally, and the serverless one when run in serverless.

I guess the only thing here to do is that maybe `boto3` should be an explicit dependency of the dagster-cloud package.

Thanks for your help @prha !

**Reply by prha** [reply]
Oh, great.  Glad you figured it out!  I think our documentation suggests that you have `dagster-cloud[serverless]` in your dependencies, which includes `boto3` (should have spotted that earlier).

**Reply by gnilrets** [reply]
Got it, I suppose this is just the pain of being an early adopter when the product continues to evolve....
